byway/src/main/webapp/routepage.js

```diff
@@ -49,27 +63,145 @@ function calcRoute(directionsService, directionsRenderer, start, end) {
   });
 }
 
-function addToStops(stop){
-  console.log("add" + stop.placename + "to the ArrayList in the servlet");
+/** Clear the stops panel in the html */
+function clearStops(){
+  const stopList = document.getElementById('stop-list');
+  if(stopList != null){
+    stopList.innerText = ""; // clear list
+  }
 }
 
-function getRecs() {
-  fetch('/api/recs')
+/** Render stop list  */
+function renderStops(stop, fromDatastore){
+  const stopList = document.getElementById('stop-list');
+  // add to stop array in the js 
+  if(fromDatastore){
+    stops.push(stop);
+  }
+  let btn = document.createElement('button');
+  btn.id = stop.id;
+  btn.innerText = stop.placename;
+  btn.className =  "btn rec-btn";
+  btn.addEventListener("click", function() {
+    deleteFromStops(stop); 
+  });
+  stopList.appendChild(btn);
+}
+
+/** Get the new list of stops from datastore onload */
+function getStopsOnload(){
+  clearStops();
+  fetch('/api/stop')
   .then(response => response.json())
-  .then((recs) => {
-    const recList = document.getElementById('rec-list');
-    recList.innerText = ""; // clear old comments 
-    recs.forEach((stop) => {
-      let btn = document.createElement("button");
-      btn.innerText = stop.placename;
-      btn.setAttribute("class", "btn rec-btn");
-      btn.addEventListener("click", function() {
-        addToStops(stop);
-      });
-      recList.appendChild(btn);
+  .then((stopsResponse) => {
+    stopsResponse.forEach((stop)=>{
+      renderStops(stop, true);
     });
+  })
+}
+
+/** Get the new list of stops locally */
+function getStopsList(){
+  clearStops();
+  // re-render list synchronously
+  stops.forEach((stop)=>{
+    renderStops(stop, false);
+  })
+}
+
+/** Add stop locally and to datastore */
+function addToStops(stop){
+  deleteFromRecs(stop); 
+  // add to stops array locally in js
+  stops.push(stop);
+  getStopsList();
+
+  // add to datastore
+  const params = new URLSearchParams();
+  params.append("text", stop.placename);
+  params.append("action", "add");
```

Comment:
Quick note: Encoding logic like this through a url parameter is not a good design pattern. It would be better to have a separate endpoint, or to make a different type of request. 

A pattern you will see more common is instead of specifying how the data is changing, you just pass the whole updated object. E.g. you remove or add what you want, and the server just replaces the whole object with whatever you provide.

Reply:
Yeah... it looked janky but it worked. I will switch to your suggested method of updating/sending the entire list instead as that mitigates this issue.